libcxx/include/__functional/move_only_function_impl.h

```diff
@@ -0,0 +1,400 @@
+    _ArgTypes...) _LIBCPP_MOVE_ONLY_FUNCTION_CV_REF noexcept(_LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT)> {
+  using _Store = __move_only_function_storage;
+
+  template <class _Functor>
+  struct __function_wrappers_impl {
+    __function_wrappers_impl() = delete;
+    __function_wrappers_impl(const __function_wrappers_impl&) = delete;
+    __function_wrappers_impl(__function_wrappers_impl&&)      = delete;
+
+    _LIBCPP_HIDE_FROM_ABI static void __destroy(void* __store) noexcept {
+      std::destroy_at(static_cast<_Functor*>(__store));
+    }
+
+    _LIBCPP_HIDE_FROM_ABI static _ReturnT
+    __call(void* __functor, _ArgTypes&&... __args) noexcept(_LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT) {
+      return std::invoke(static_cast<_Functor _LIBCPP_MOVE_ONLY_FUNCTION_INV_QUALS>(*static_cast<_Functor*>(__functor)),
+                         std::forward<_ArgTypes>(__args)...);
+    }
+  };
+
+  template <class _Functor>
+  using __function_wrappers = __function_wrappers_impl<remove_cvref_t<_Functor>>;
+
+  template <class _VT>
+  consteval static bool __is_callable_from_impl() {
+    if (_LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT) {
+      return is_nothrow_invocable_r_v< _ReturnT, _VT _LIBCPP_MOVE_ONLY_FUNCTION_CV_REF, _ArgTypes...> &&
+             is_nothrow_invocable_r_v<_ReturnT, _VT _LIBCPP_MOVE_ONLY_FUNCTION_INV_QUALS, _ArgTypes...>;
+    } else {
+      return is_invocable_r_v<_ReturnT, _VT _LIBCPP_MOVE_ONLY_FUNCTION_CV_REF, _ArgTypes...> &&
+             is_invocable_r_v<_ReturnT, _VT _LIBCPP_MOVE_ONLY_FUNCTION_INV_QUALS, _ArgTypes...>;
+    }
+    return false;
+  }
+
+  template <class _VT>
+  static constexpr bool __is_callable_from = __is_callable_from_impl<decay_t<_VT>>();
+
+  template <class _Func, class... _Args>
+  _LIBCPP_HIDE_FROM_ABI void __construct(_Args&&... __args) {
+    static_assert(is_constructible_v<decay_t<_Func>, _Func>);
+
+    using _FuncWraps = __function_wrappers<_Func>;
+    using _UnRefFunc = remove_reference_t<_Func>;
+
+    if constexpr (_Store::__fits_in_trivially_destructible_buffer<_Func>) {
+      auto& __data   = __storage_.__trivially_destructible_;
+      __data.__call_ = std::bit_cast<void*>(&_FuncWraps::__call);
+      std::construct_at(reinterpret_cast<_UnRefFunc*>(__data.__data_), std::forward<_Args>(__args)...);
+      __data.__status_ = _Store::_Status::_TriviallyDestructible;
+    } else if constexpr (_Store::__fits_in_trivially_relocatable_buffer<_Func>) {
+      auto& __data      = __storage_.__trivially_relocatable_;
+      __data.__call_    = std::bit_cast<void*>(&_FuncWraps::__call);
+      __data.__destroy_ = &_FuncWraps::__destroy;
+      std::construct_at(reinterpret_cast<_UnRefFunc*>(__data.__data_), std::forward<_Args>(__args)...);
+      __data.__status_ = _Store::_Status::_TriviallyRelocatable;
+    } else {
+      auto& __data      = __storage_.__large_;
+      __data.__call_    = std::bit_cast<void*>(&_FuncWraps::__call);
+      __data.__destroy_ = &_FuncWraps::__destroy;
```

Comment:
I don't see the reason to indirect/reuse-the-name `_FuncWraps::__destroy` on lines 242 and 248. We're only going to instantiate _one_ of those two lines for any given type `_Func`, so you might as well use a lambda. Here, for example,
```
__data.__call_ = static_cast<void(*)()>(+[](void *__p, _ArgTypes&&... __args) {
  return std::invoke(static_cast<_Func _LIBCPP_MOVE_ONLY_FUNCTION_INV_QUALS>(*static_cast<_Func*>(__p)),
                     std::forward<_ArgTypes>(__args)...);
});
__data.__destroy_ = +[](void *__p) {
  std::destroy_at(static_cast<_Functor*>(__p));
};
```
And then, because this is actually the `_Heap` case, we can regularize our lambdas to operate on the stored pointer instead of forcing our caller to pass in _the thing the pointer points to_:
```
__data.__call_ = static_cast<void(*)()>(+[](void *__store, _ArgTypes&&... __args) {
  void *__p = *(void**)__store;
  return std::invoke(static_cast<_Func _LIBCPP_MOVE_ONLY_FUNCTION_INV_QUALS>(*static_cast<_Func*>(__p)),
                     std::forward<_ArgTypes>(__args)...);
});
__data.__destroy_ = +[](void *__store) {
  void *__p = *(void**)__store;
  std::destroy_at(static_cast<_Functor*>(__p));
  ::operator delete(__p);
};
```
See my next comment, on function `__reset`.